### ISSUE TITLE
Hyperoptimize uses far less GPU memory than expected, and is slower than running standalone models

### ISSUE THREAD

**Issue by erikr** [aguirre-lab/ml4c3#1335]
## Error message or warning
`nvidia-smi` output when training via `hyperoptimize` to GPUs 0 and 1 indicates only 312MB of GPU memory are being used:

```console
Tue Feb  9 08:18:28 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Quadro RTX 8000     Off  | 00000000:19:00.0 Off |                  Off |
| 33%   33C    P8    21W / 260W |    312MiB / 48601MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Quadro RTX 8000     Off  | 00000000:1A:00.0 Off |                  Off |
| 34%   39C    P8    27W / 260W |    312MiB / 48601MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Quadro RTX 8000     Off  | 00000000:67:00.0 Off |                  Off |
| 34%   50C    P0    80W / 260W |      0MiB / 48601MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Quadro RTX 8000     Off  | 00000000:68:00.0 Off |                  Off |
| 33%   50C    P0    48W / 260W |      0MiB / 48598MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    265331      C   ...sts-ecg/hyperoptimize/v01      309MiB |
|    1   N/A  N/A    265338      C   ...sts-ecg/hyperoptimize/v01      309MiB |
+-----------------------------------------------------------------------------+
```

However, running one of these models on GPU 0 "manually" from the command line consumes much more GPU memory and completes much faster:
```console
Tue Feb  9 08:20:10 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Quadro RTX 8000     Off  | 00000000:19:00.0 Off |                  Off |
| 33%   37C    P2    61W / 260W |   7366MiB / 48601MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Quadro RTX 8000     Off  | 00000000:1A:00.0 Off |                  Off |
| 33%   41C    P0    70W / 260W |      0MiB / 48601MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Quadro RTX 8000     Off  | 00000000:67:00.0 Off |                  Off |
| 34%   49C    P0    79W / 260W |      0MiB / 48601MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Quadro RTX 8000     Off  | 00000000:68:00.0 Off |                  Off |
| 33%   49C    P0    48W / 260W |      0MiB / 48598MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    266613      C   ...ults/ecgnet-v011-newest/0     7363MiB |
+-----------------------------------------------------------------------------+
```

## Expected behavior
Training a model on a GPU via `hyperoptimize` uses the same amount of memory as if one trained that model standalone and not in `hyperoptimize`, and completes in the same amount of time.

## Steps to reproduce
```bash
PREFIX=$HOME/dropbox/sts-data
BOOTSTRAP=0

./scripts/run.sh -d 0,1 $PWD/ml4c3/recipes.py hyperoptimize \
--max_evals 4 \
--hyperoptimize_config_file $HOME/dropbox/sts-ecg/scripts/hyperoptimize-ecgnet-recipe-optimizer-schedule.json \
--num_workers 4 \
--tensors     /storage/shared/ecg/mgh \
--tensors     $PREFIX/mgh.csv sts \
--train_csv   $PREFIX/bootstraps-intersect-ecg-age-sex-metadata-newest/$BOOTSTRAP/train.csv \
--valid_csv   $PREFIX/bootstraps-intersect-ecg-age-sex-metadata-newest/$BOOTSTRAP/valid.csv \
--test_csv    $PREFIX/bootstraps-intersect-ecg-age-sex-metadata-newest/$BOOTSTRAP/test.csv \
--input_tensors \
    ecg_2500_zscore_pop_60_days_before_sts_newest \
--output_tensors \
    sts_death_60_days_after_ecg_newest \
    sts_stroke_60_days_after_ecg_newest \
    sts_renal_failure_60_days_after_ecg_newest \
    sts_prolonged_ventilation_60_days_after_ecg_newest \
    sts_dsw_infection_60_days_after_ecg_newest \
    sts_reoperation_60_days_after_ecg_newest \
    sts_long_stay_60_days_after_ecg_newest \
--conv_x 30 \
--conv_blocks 32 \
--conv_block_size 1 \
--conv_block_layer_order convolution normalization activation dropout \
--dense_blocks 10 20 40 80 160 \
--dense_block_size 2 \
--dense_block_layer_order convolution normalization activation dropout \
--dense_layers 32 32 \
--pool_x 2 \
--pool_type max \
--bottleneck_type flatten_restructure \
--activation_layer relu \
--spatial_dropout 0.25 \
--dense_dropout 0.5 \
--optimizer adam \
--learning_rate 0.0001 \
--learning_rate_patience 8 \
--learning_rate_reduction 0.5 \
--patience 24 \
--epochs 256 \
--batch_size 256 \
--output_folder $HOME/sts-ecg/hyperoptimize/v01
```

**Comment by erikr** [comment]
@StevenSong Can you take a look? I'll own https://github.com/aguirre-lab/ml4c3/issues/1325.

**Comment by StevenSong** [comment]
Debugged this a bit, I should have looked in the config `.json` file when you asked me on slack yesterday:

> However, running one of these models on GPU 0 "manually" from the command line consumes much more GPU memory and completes much faster:

did you try the hyperparameters from the config file in the manual test? when I run it with `adamw` or `learning_rate_schedule`, I get errors from the manual run.

the error with `learning_rate_schedule` is 2 fold and I do remember when we were ripping things out from our fork, I disabled alot of things that broke learning rate schedule. we decided then to leave it until we wanted to use it - perhaps now is that time?

additionally, `adamw` doesnt seem to be an optimizer that is available, at least in my setup, did you perhaps build a new docker image with `adamw` installed?

I find it weird though that hyperopt wasnt killing these trials which would error, that is another issue to solve

so tl;dr, we need to:
1. fix `learning_rate_schedule`
2. install `adamw`
3. make `hyperopt` recipe robust to trials failing

a last question, what's the thought behind these experiments and results for paper?

**Comment by StevenSong** [comment]
Oh i see, you've created some of these issues already, let me try from your branch again

**Comment by erikr** [comment]
My bad, forgot to add `git checkout er-adamw` to the `Steps to reproduce`!

**Comment by erikr** [comment]
I am able to run the following optimizers in standalone `train` runs successfully on my branch:
```
sgd
adam
adamw
radam
```

**Comment by StevenSong** [comment]
> I am able to run the following optimizers in standalone `train` runs successfully on my branch:

were u also able to use learning_rate_schedule with these?

**Comment by StevenSong** [comment]
I'm unable to run it with learning rate schedule - it looks just like some minor bugs in the implementation #1334, i'll leave comments there

**Comment by erikr** [comment]
> I'm unable to run it with learning rate schedule - it looks just like some minor bugs in the implementation #1334, i'll leave comments there

I am able to run it for all optimizers.

I presume you are encountering this error https://github.com/aguirre-lab/ml4c3/blob/cc31849535222cf0fe274f740e32b7c3aef208cb/ml4c3/arguments.py#L1550-L1555

Just set `patience > epochs` as the error message states and it will run.

**Comment by erikr** [comment]
I cannot reproduce this `hyperoptimize` issue on `master`, so I think this problem is related to bugs in the implementation of `adamw`, `radam`, and cyclical learning rate schedulers.

**Comment by StevenSong** [comment]
lots of debugging on slack, the tl;dr:
1. need to use keyword `learning_rate` rather than `lr`
2. that introduces another error which is fixed in TFA 0.12 which requires TF 2.4 - so we'd need to upgrade (https://github.com/tensorflow/addons/pull/2172)

**Comment by erikr** [comment]
It is working now!
```
Tue Feb  9 21:01:18 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce RTX 208...  Off  | 00000000:19:00.0 Off |                  N/A |
| 68%   75C    P2   239W / 250W |   5544MiB / 11019MiB |     96%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  GeForce RTX 208...  Off  | 00000000:1A:00.0 Off |                  N/A |
| 91%   80C    P2    92W / 250W |   7064MiB / 11019MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  GeForce RTX 208...  Off  | 00000000:67:00.0 Off |                  N/A |
| 91%   88C    P2   236W / 250W |   5528MiB / 11019MiB |     91%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  GeForce RTX 208...  Off  | 00000000:68:00.0 Off |                  N/A |
| 93%   88C    P2   141W / 250W |  10994MiB / 11018MiB |     98%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    999726      C   ...sts-ecg/hyperoptimize/v01     5541MiB |
|    1   N/A  N/A    999733      C   ...sts-ecg/hyperoptimize/v01     7061MiB |
|    2   N/A  N/A    999740      C   ...sts-ecg/hyperoptimize/v01     5525MiB |
|    3   N/A  N/A   1001587      C   ...sts-ecg/hyperoptimize/v01    10991MiB |
+-----------------------------------------------------------------------------+
```

**Comment by erikr** [comment]
Now it seems to allocate multiple models to a GPU even if a model is already running on it. Note how the memory of each GPU is near max capacity:
```
Wed Feb 10 09:18:29 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce RTX 208...  Off  | 00000000:19:00.0 Off |                  N/A |
|  0%   44C    P8    16W / 250W |  10994MiB / 11019MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  GeForce RTX 208...  Off  | 00000000:1A:00.0 Off |                  N/A |
| 88%   85C    P2   232W / 250W |   9624MiB / 11019MiB |     99%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  GeForce RTX 208...  Off  | 00000000:67:00.0 Off |                  N/A |
| 68%   65C    P8    36W / 250W |  10136MiB / 11019MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  GeForce RTX 208...  Off  | 00000000:68:00.0 Off |                  N/A |
| 93%   87C    P2   154W / 250W |  10994MiB / 11018MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   1161345      C   ...sts-ecg/hyperoptimize/v01    10991MiB |
|    1   N/A  N/A   1155939      C   ...sts-ecg/hyperoptimize/v01     9621MiB |
|    2   N/A  N/A   1181525      C   ...sts-ecg/hyperoptimize/v01    10133MiB |
```

and TensorFlow is telling us:
```
2021-02-10 09:18:00.659397: I tensorflow/core/common_runtime/bfc_allocator.cc:1040] Sum Total of in-use chunks: 6.95GiB
2021-02-10 09:18:00.659408: I tensorflow/core/common_runtime/bfc_allocator.cc:1042] total_region_allocated_bytes_: 9662627840 memory_limit_: 10561793600 available bytes: 899165760 curr_region_allocation_bytes_: 8589934592
2021-02-10 09:18:00.659431: I tensorflow/core/common_runtime/bfc_allocator.cc:1048] Stats:
Limit:                     10561793600
InUse:                      7467591168
MaxInUse:                   8491591168
NumAllocs:                         632
MaxAllocSize:               3856777216
Reserved:                            0
PeakReserved:                        0
LargestFreeBlock:                    0

2021-02-10 09:18:00.659514: W tensorflow/core/common_runtime/bfc_allocator.cc:441] ***___*****_____***********************_____****************_********************__*****************
2021-02-10 09:18:00.659560: W tensorflow/core/framework/op_kernel.cc:1763] OP_REQUIRES failed at transpose_op.cc:184 : Resource exhausted: OOM when allocating tensor with shape[256,400,1,2500] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
```

**Comment by StevenSong** [comment]
> Now it seems to allocate multiple models to a GPU even if a model is already running on it. Note how the memory of each GPU is near max capacity:

I don't know if this means that there are multiple models running, the logic of hyperopt doesnt really allow for that (each worker is dispatched with a GPU and the GPU wont be re-dispatched until the last job on that GPU is finished). Is there something that makes you suspect there actually is more than 1 model per gpu? going off just the memory usage, the usage of 10991 MiB is the same as what was running on GPU 3 last night - maybe some models are just much much bigger?

**Comment by erikr** [comment]
I have not confirmed >1 model per GPU, just an assumption when so much memory is in use. But some of the hyperparameter configurations result in large models, e.g. many layers with many neurons. So it is possible that some models are just bigger.
